pkg/kube-metrics/collector.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"context"
+
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	"k8s.io/kube-state-metrics/pkg/metric"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+)
+
+// NewCollectors returns collections of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+// Current operators namespace will be added to the passed namespaces if it is not present.
+func NewCollectors(uc *Client, namespaces []string, api string, kind string, metricFamily []metric.FamilyGenerator) ([]*kcollector.Collector, error) {
+	// Detect in which namespace the operator is deployed in.
+	operatorNamespace, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		return nil, err
+	}
+	namespaces = append(namespaces, operatorNamespace)
+	namespaces = deduplicateNamespaces(namespaces)
+	var collectors []*kcollector.Collector
+	// Generate collector per namespace.
+	for _, ns := range namespaces {
```

Comment:
I am worried that this will not find all resources if an operator is watching the entire cluster or a series of namespaces. I think that the ns list is empty until we add just the NS that the operator is installed in.

Am I misunderstanding how this works?

Reply:
We pass in the `namespaces` list which is provided by the user, so the user can always pass in any needed namespace by getting them from the API. By default, we detect the namespace the operator is running in and append that to the list. We deduplicate the namespaces, as otherwise we get duplicated metrics.

Reply:
We could add a helper function the user can use to detect all namespaces?

Reply:
I am worried about the case where a namespace is added after the operator is running. I think that we can deal with this as a follow-up though.

Reply:
@shawn-hurley Agreed will open issue once this is merged! Thanks!